onnx/printer.py

```diff
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import onnx
+import onnx.onnx_cpp2py_export.printer as C
+from typing import Text, Union
+
+
+def to_text(proto: Union[onnx.ModelProto, onnx.FunctionProto, onnx.GraphProto]) -> Text:
+    if isinstance(proto, onnx.ModelProto):
```

Comment:
It is better this way. There should be a documentation for this function. I still think the call to SerializeToString() should be avoided.

Reply:
How do we avoid the serialization? (Of course, we can reimplement the pretty-printer in Python, but I assume that's not what you mean? That would lead to a potential divergence of behavior between C++ and Python. Since this usage is mostly for debugging and understanding etc., I think the efficiency issue is not that critical.)

Reply:
The model is serialized here and then deserialized in C++. I think these steps should be avoided. The only change to make that happen is to change the C++ signature to take FunctionProto, ModelProto or a GraphProto and remove the call to ParseFromString. Users may call this function on deep learning models to compare two graphs. Avoiding serialization could make it a lot faster.

Reply:
Specifically, I don't think that Python serialization and C++ deserialization are inverses of each other (at least, it is not documented to be so, and I doubt it is). So, their composition is not a "no-op", but it serves to convert from the python-representation to C++-representation.

Reply:
A Python ModelProto and a C++ ModelProto are different: the Python ModelProto wraps the C++ ModelProto, probably in a capsule. I expect that the Python and C++ SerializeToString to produce the exact same sequence of bytes. That's what pybind11 does. They use swig to do it for ModelProto. It should work in a similar way.

Reply:
I did some measurements. I was hoping to see that serialization is independant of the size but it does not. So call SerializeToString and then ParseFromString is really inefficient.

![Figure_1](https://user-images.githubusercontent.com/22452781/173557024-2ebff163-2dac-4c33-aeb7-0461daf05587.png)

Reply:
> A Python ModelProto and a C++ ModelProto are different: the Python ModelProto wraps the C++ ModelProto, probably in a capsule. I expect that the Python and C++ SerializeToString to produce the exact same sequence of bytes. That's what pybind11 does. They use swig to do it for ModelProto. It should work in a similar way.

Of course, SerializeToString will produce the same sequence of bytes, since that is defined by the Protobuf semantics. But that is not the question. The question is whether the C++ in-memory-representation is available in a Python ModelProto (as you imply). If so, what is the API/call we can use to get the C++ in-memory-representation from a Python ModelProto?

Reply:
(The question is about the Deserialize part, that is: creating the in-memory-representation used by the two languages).

Reply:
See https://github.com/onnx/onnx/pull/4274.